src/ripple/app/tx/impl/AMMDeposit.cpp

```diff
@@ -0,0 +1,528 @@
+    // as we go on processing transactions.
+    Sandbox sb(&ctx_.view());
+
+    // This is a ledger with just the fees paid and any unfunded or expired
+    // offers we encounter removed. It's used when handling Fill-or-Kill offers,
+    // if the order isn't going to be placed, to avoid wasting the work we did.
+    Sandbox sbCancel(&ctx_.view());
+
+    auto const result = applyGuts(sb);
+    if (result.second)
+        sb.apply(ctx_.rawView());
+    else
+        sbCancel.apply(ctx_.rawView());
+
+    return result.first;
+}
+
+std::pair<TER, STAmount>
+AMMDeposit::deposit(
+    Sandbox& view,
+    AccountID const& ammAccount,
+    STAmount const& asset1Deposit,
+    std::optional<STAmount> const& asset2Deposit,
+    STAmount const& lpTokensDeposit)
+{
+    // Check account has sufficient funds
+    auto balance = [&](auto const& asset) {
+        return accountHolds(
+                   view,
+                   account_,
+                   asset.issue().currency,
+                   asset.issue().account,
+                   FreezeHandling::fhZERO_IF_FROZEN,
+                   ctx_.journal) >= asset;
+    };
+
+    // Deposit asset1Deposit
+    if (!balance(asset1Deposit))
+    {
+        JLOG(ctx_.journal.debug())
+            << "AMM Deposit: account has insufficient balance to deposit "
+            << asset1Deposit;
+        return {tecUNFUNDED_AMM, STAmount{}};
+    }
+    auto res = ammSend(view, account_, ammAccount, asset1Deposit, ctx_.journal);
+    if (res != tesSUCCESS)
+    {
+        JLOG(ctx_.journal.debug())
+            << "AMM Deposit: failed to deposit " << asset1Deposit;
+        return {res, STAmount{}};
+    }
+
+    // Deposit asset2Deposit
+    if (asset2Deposit)
+    {
+        if (!balance(*asset2Deposit))
+        {
+            JLOG(ctx_.journal.debug())
+                << "AMM Deposit: account has insufficient balance to deposit "
+                << *asset2Deposit;
+            return {tecUNFUNDED_AMM, STAmount{}};
+        }
+        res = ammSend(view, account_, ammAccount, *asset2Deposit, ctx_.journal);
+        if (res != tesSUCCESS)
+        {
+            JLOG(ctx_.journal.debug())
+                << "AMM Deposit: failed to deposit " << *asset2Deposit;
+            return {res, STAmount{}};
+        }
+    }
+
+    // Deposit LP tokens
+    res =
+        accountSend(view, ammAccount, account_, lpTokensDeposit, ctx_.journal);
+    if (res != tesSUCCESS)
+    {
+        JLOG(ctx_.journal.debug()) << "AMM Deposit: failed to deposit LPTokens";
+        return {res, STAmount{}};
+    }
+
+    return {tesSUCCESS, lpTokensDeposit};
+}
+
+/** Proportional deposit of pools assets in exchange for the specified
+ * amount of LPTokens.
+ */
+std::pair<TER, STAmount>
+AMMDeposit::equalDepositTokens(
+    Sandbox& view,
+    AccountID const& ammAccount,
+    STAmount const& asset1Balance,
+    STAmount const& asset2Balance,
+    STAmount const& lptAMMBalance,
+    STAmount const& lpTokensDeposit)
+{
```

Comment:
These functions can throw. Doe they need to be wrapped in a try/catch and the appropriate error code returned. Or if they don't need to be wrapped, then a comment explaining why not should be added.

Reply:
Thanks. Added here and in AMMWithdraw::equalWithdrawTokens().